### PR TITLE
add fuzzer for Allocator

### PR DESCRIPTION
**Pull request by arvidn**
specifically handling of `atom()`, `small_number()` and `number()`.

This checks certain properties that we expect to hold for any atom (conditioned on whether it's an canonical form integer or not).

This functions as a property-based test.

### CODE REVIEW

File: src/allocator.rs [hunk 1]
```diff
@@ -1082,7 +1082,9 @@ fn test_concat_launder_small_number() {
 
     // this "launders" the small number into actually being allocated on the
     // heap
-    let atom2 = a.new_concat(1, &[atom1]).expect("new_substr");
+    let atom2 = a
+        .new_concat(1, &[a.nil(), atom1, a.nil()])
```
Comment (arvidn) [inline src/allocator.rs:1086]: this is somewhat unrelated. it makes it more likely that this `concat` actually causes a heap allocation for the return value

**Review by Rigidity — APPROVED** [review]
lgtm
